File: spine-cocos2dxna/src/spine/CCSkeleton.cs
/******************************************************************************
 * Spine Runtimes Software License
 * Version 2
 * 
 * Copyright (c) 2013, Esoteric Software
 * All rights reserved.
 * 
 * You are granted a perpetual, non-exclusive, non-sublicensable and
 * non-transferable license to install, execute and perform the Spine Runtimes
 * Software (the "Software") solely for internal use. Without the written
 * permission of Esoteric Software, you may not (a) modify, translate, adapt or
 * otherwise create derivative works, improvements of the Software or develop
 * new applications using the Software or (b) remove, delete, alter or obscure
 * any trademarks or any copyright, trademark, patent or other intellectual
 * property or proprietary rights notices on or in the Software, including
 * any copy thereof. Redistributions in binary or source form must include
 * this license and terms. THIS SOFTWARE IS PROVIDED BY ESOTERIC SOFTWARE
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL ESOTERIC SOFTARE BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *****************************************************************************/

using Cocos2D;
using Spine;
using SpineCocosXna.Spine.Data.Cocos;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SpineCocosXna.Spine
{

    public enum spAttachmentType
    {
        ATTACHMENT_REGION = 1, ATTACHMENT_REGION_SEQUENCE = 2, ATTACHMENT_BOUNDING_BOX = 3
    }

    public class CCSkeleton : CCNodeRGBA
    {

        public float FLT_MAX = 3.402823466e+38F;     /* max value */
        public float FLT_MIN = 1.175494351e-38F;     /* min positive value */

        public static int ATTACHMENT_REGION = 0;
        public static int ATTACHMENT_REGION_SEQUENCE = 1;
        public static int ATTACHMENT_BOUNDING_BOX = 2;

        public CCBlendFunc blendFunc = new CCBlendFunc();

        public Skeleton skeleton;
        public Bone rootBone;

        float timeScale;
        bool debugSlots = true;
        bool debugBones;
        bool premultipliedAlpha;

        bool ownsSkeletonData;
        Atlas atlas;

        static CCSkeleton createWithData(SkeletonData skeletonData, bool ownsSkeletonData = false)
        {
            CCSkeleton node = new CCSkeleton(skeletonData, ownsSkeletonData);
            return node;
        }

        static CCSkeleton createWithFile(string skeletonDataFile, Atlas atlas, float scale = 0)
        {
            CCSkeleton node = new CCSkeleton(skeletonDataFile, atlas, scale);
            return node;
        }

        static CCSkeleton createWithFile(string skeletonDataFile, string atlasFile, float scale = 0)
        {
            CCSkeleton node = new CCSkeleton(skeletonDataFile, atlasFile, scale);
            return node;
        }

        public void initialize()
        {
            atlas = null;
            debugSlots = false;
            debugBones = false;
            timeScale = 1;

            blendFunc = new CCBlendFunc(CCOGLES.GL_ONE, CCOGLES.GL_ONE_MINUS_SRC_ALPHA);
            CCDrawManager.BlendFunc(blendFunc);

            setOpacityModifyRGB(true);
             
            //SetShaderProgram(CCShaderCache::sharedShaderCache()->programForKey(kCCShader_PositionTextureColor));

            ScheduleUpdate();
        }

        public void setSkeletonData(SkeletonData skeletonData, bool ownsSkeletonData)
        {
            skeleton = new Skeleton(skeletonData);
            rootBone = skeleton.Bones[0];
            this.ownsSkeletonData = ownsSkeletonData;
        }

        public CCSkeleton()
        {
            initialize();
        }

        public CCSkeleton(SkeletonData skeletonData, bool ownsSkeletonData = false)
        {
            initialize();
            setSkeletonData(skeletonData, ownsSkeletonData);
        }

        public CCSkeleton(string skeletonDataFile, Atlas atlas, float scale = 0)
        {
            var json = new SkeletonJson(atlas);
            json.Scale = scale == 0 ? (1 / CCDirector.SharedDirector.ContentScaleFactor) : scale;
            SkeletonData skeletonData = json.ReadSkeletonData(skeletonDataFile);
            setSkeletonData(skeletonData, true);
        }

        public CCSkeleton(string skeletonDataFile, string atlasFile, float scale = 0)
        {

            initialize();

            atlas = new Atlas(atlasFile, new MonoTextureLoader(CCDrawManager.GraphicsDevice));

            SkeletonJson json = new SkeletonJson(atlas);

            json.Scale = scale == 0 ? (1 / CCDirector.SharedDirector.ContentScaleFactor) : scale;

            SkeletonData skeletonData = json.ReadSkeletonData(skeletonDataFile);
            setSkeletonData(skeletonData, true);
        }

        ~CCSkeleton()
        {
            
            //Eliminamos el skleñetp
            //if (ownsSkeletonData!=null) 

            //SkeletonData_dispose(skeleton->data);
            //if (atlas) spAtlas_dispose(atlas);
            //spSkeleton_dispose(skeleton);
            
        }
      
        public override void Update(float dt)
        {
            base.Update(dt);
            skeleton.Update(dt*timeScale);
        }

        public override void Draw()
        {
            base.Draw();

            // CC_NODE_DRAW_SETUP();

            CCDrawManager.BlendFunc(blendFunc);
            CCColor3B color = Color;
            skeleton.R = color.R / 255f ; 
            skeleton.G = color.G / 255f; 
            skeleton.B = color.B / 255f; 
            skeleton.A = Opacity / 255f;

            bool additive = false;

            CCTextureAtlas textureAtlas = null;

            CCV3F_C4B_T2F_Quad quad = new CCV3F_C4B_T2F_Quad();

            quad.TopLeft.Vertices.Z = 0;
            quad.TopRight.Vertices.Z = 0;
            quad.BottomLeft.Vertices.Z = 0;
            quad.BottomRight.Vertices.Z = 0;

            for (int i = 0; i < skeleton.Slots.Count; i++)
            {
                Slot slot = skeleton.DrawOrder[i];

                if (slot.Attachment == null || slot.Attachment.GetType().ToString() != "Spine.RegionAttachment") continue;  //=> ??????????????


                RegionAttachment attachment = (RegionAttachment)slot.Attachment;
                CCTextureAtlas regionTextureAtlas = GetTextureAtlas(attachment);
                regionTextureAtlas.DrawQuads(); //retain();
                
                //self.rendererObject = textureAtlas;
                //self.width = texture.PixelsWide;
                //self.height = texture.PixelsHigh;

                if (slot.Data.AdditiveBlending != additive)
                {
                    if (textureAtlas != null)
                    {
                        textureAtlas.DrawQuads();
                        textureAtlas.RemoveAllQuads();
                    }

                    additive = !additive;

                    CCDrawManager.BlendFunc(new CCBlendFunc(blendFunc.Source, (additive) ? CCOGLES.GL_ONE : blendFunc.Destination) );

                }
                else if (textureAtlas != null && regionTextureAtlas != textureAtlas)
                {
                    textureAtlas.DrawQuads();
                    textureAtlas.RemoveAllQuads();
                }

                textureAtlas = regionTextureAtlas;

                int quadCount = textureAtlas.TotalQuads;
                if (textureAtlas.Capacity == quadCount)
                {
                    textureAtlas.DrawQuads();
                    textureAtlas.RemoveAllQuads();

                    if (!textureAtlas.ResizeCapacity(textureAtlas.Capacity * 2)) return;
                }
                
                spRegionAttachment_updateQuad(attachment, slot, ref quad, premultipliedAlpha);

                textureAtlas.UpdateQuad(ref quad, quadCount);

            }
          
            if (textureAtlas != null)
            {
                textureAtlas.DrawQuads();
                textureAtlas.RemoveAllQuads();
            }

           
            if (debugSlots)
            {

                CCPoint[] points = new CCPoint[4];

                CCV3F_C4B_T2F_Quad quada = new CCV3F_C4B_T2F_Quad();

                for (int i = 0, n = skeleton.Slots.Count; i < n; i++)
                {

                    Slot slot = skeleton.DrawOrder[i];

                    if (slot.Attachment == null || slot.Attachment.Name != "ATTACHMENT_REGION") continue;

                    RegionAttachment attachment = (RegionAttachment)slot.Attachment;
                    spRegionAttachment_updateQuad(attachment, slot, ref quad);

                    points[0] = new CCPoint(quada.BottomLeft.Vertices.X, quada.BottomLeft.Vertices.Y);
                    points[1] = new CCPoint(quada.BottomRight.Vertices.X, quada.BottomRight.Vertices.Y);
                    points[2] = new CCPoint(quada.TopRight.Vertices.X, quada.TopRight.Vertices.Y);
                    points[3] = new CCPoint(quada.TopLeft.Vertices.X, quada.TopLeft.Vertices.Y);

                    CCDrawingPrimitives.DrawPoly(points, 4, true, new CCColor4B(0, 0, 255, 255));
                }
            }

            if (debugBones)
            {
                // Bone lengths.
                for (int i = 0; i < skeleton.Bones.Count; i++)
                {
                    Bone bone = skeleton.Bones[i];
                    float x = bone.Data.Length * bone.M00 + bone.WorldX;
                    float y = bone.Data.Length * bone.M10 + bone.WorldY;

                    CCDrawingPrimitives.Begin();
                    CCDrawingPrimitives.DrawLine(new CCPoint(bone.WorldX, bone.WorldY), new CCPoint(x, y), new CCColor4B(255, 0, 0, 255));
                    CCDrawingPrimitives.End();
                }

                // Bone origins.
                for (int i = 0; i < skeleton.Bones.Count; i++)
                {
                    Bone bone = skeleton.Bones[i];
                    CCDrawingPrimitives.Begin();
                    CCDrawingPrimitives.DrawPoint(new CCPoint(bone.WorldX, bone.WorldY), 4, new CCColor4B(0, 0, 255, 255));
                    CCDrawingPrimitives.End();

                }
            }

        }


        public CCTextureAtlas GetTextureAtlas(RegionAttachment attachment)
        {
            AtlasRegion regionAtlas = (AtlasRegion)attachment.RendererObject;
            var texture2D = (Microsoft.Xna.Framework.Graphics.Texture2D)regionAtlas.page.rendererObject;
            //Console.WriteLine(texture2D.IsDisposed);
           
            return spine_cocos2dx.CreateAtlasFromTexture2D(texture2D);
        }

        public CCRect boundingBox()
        {
            float minX = FLT_MAX, minY = FLT_MAX, maxX = FLT_MIN, maxY = FLT_MIN;


            float[] vertices = new float[8];


            for (int i = 0; i < skeleton.Slots.Count; ++i)
            {

                Slot slot = skeleton.Slots[i];

                if (slot.Attachment != null || slot.Attachment.Name != "ATTACHMENT_REGION") continue;

                RegionAttachment attachment = (RegionAttachment)slot.Attachment;

                attachment.ComputeWorldVertices(slot.Skeleton.X, slot.Skeleton.Y, slot.Bone, vertices);

                minX = Math.Min(minX, vertices[RegionAttachment.X1] * ScaleX);
                minY = Math.Min(minY, vertices[RegionAttachment.Y1] * ScaleY);
                maxX = Math.Max(maxX, vertices[RegionAttachment.X1] * ScaleX);
                maxY = Math.Max(maxY, vertices[RegionAttachment.Y1] * ScaleY);
                minX = Math.Min(minX, vertices[RegionAttachment.X4] * ScaleX);
                minY = Math.Min(minY, vertices[RegionAttachment.Y4] * ScaleY);
                maxX = Math.Max(maxX, vertices[RegionAttachment.X4] * ScaleX);
                maxY = Math.Max(maxY, vertices[RegionAttachment.Y4] * ScaleY);
                minX = Math.Min(minX, vertices[RegionAttachment.X2] * ScaleX);
                minY = Math.Min(minY, vertices[RegionAttachment.Y2] * ScaleY);
                maxX = Math.Max(maxX, vertices[RegionAttachment.X2] * ScaleX);
                maxY = Math.Max(maxY, vertices[RegionAttachment.Y2] * ScaleY);
                minX = Math.Min(minX, vertices[RegionAttachment.X3] * ScaleX);
                minY = Math.Min(minY, vertices[RegionAttachment.Y3] * ScaleY);
                maxX = Math.Max(maxX, vertices[RegionAttachment.X3] * ScaleX);
                maxY = Math.Max(maxY, vertices[RegionAttachment.Y3] * ScaleY);
            }
            CCPoint position = Position;

            return new CCRect(position.X + minX, position.Y + minY, maxX - minX, maxY - minY);

        }

        // --- Convenience methods for common Skeleton_* functions.
        public void updateWorldTransform()
        {
            skeleton.UpdateWorldTransform();
        }

        public void setToSetupPose()
        {
            skeleton.SetToSetupPose();
        }
        public void setBonesToSetupPose()
        {
            skeleton.SetBonesToSetupPose();
        }
        public void setSlotsToSetupPose()
        {
            skeleton.SetSlotsToSetupPose();
        }

        /* Returns 0 if the bone was not found. */
        public Bone findBone(string boneName)
        {
            return skeleton.FindBone(boneName);
        }
        /* Returns 0 if the slot was not found. */
        public Slot findSlot(string slotName)
        {
            return skeleton.FindSlot(slotName);
        }

        /* Sets the skin used to look up attachments not found in the SkeletonData defaultSkin. Attachments from the new skin are
         * attached if the corresponding attachment from the old skin was attached. Returns false if the skin was not found.
         * @param skin May be 0.*/
        public bool setSkin(string skinName)
        {
            skeleton.SetSkin(skinName);
            return true;
        }

        /* Returns 0 if the slot or attachment was not found. */
        public Attachment getAttachment(string slotName, string attachmentName)
        {
            return skeleton.GetAttachment(slotName, attachmentName);
        }
        /* Returns false if the slot or attachment was not found. */
        public bool setAttachment(string slotName, string attachmentName)
        {
            skeleton.SetAttachment(slotName, attachmentName);
            return true;
        }

        // --- CCBlendProtocol
        //CC_PROPERTY(cocos2d::ccBlendFunc, blendFunc, BlendFunc);
        public void setOpacityModifyRGB(bool value)
        {
            premultipliedAlpha = value;
        }

        public bool isOpacityModifyRGB()
        {
            return premultipliedAlpha;
        }
      
        //virtual cocos2d::CCTextureAtlas* getTextureAtlas (RegionAttachment regionAttachment);
        #region SpinesCocos2d

        void spRegionAttachment_updateQuad(RegionAttachment self, Slot slot, ref CCV3F_C4B_T2F_Quad quad, bool premultipliedAlpha = false)
        {

            float[] vertices = new float[8];

            self.ComputeWorldVertices(slot.Skeleton.X, slot.Skeleton.Y, slot.Bone, vertices);

            float r = slot.Skeleton.R * slot.R * 255;
            float g = slot.Skeleton.G * slot.G * 255;
            float b = slot.Skeleton.B * slot.B * 255;

            float normalizedAlpha = slot.Skeleton.A * slot.A;
            if (premultipliedAlpha)
            {
                r *= normalizedAlpha;
                g *= normalizedAlpha;
                b *= normalizedAlpha;
            }

            float a = normalizedAlpha * 255;
            quad.BottomLeft.Colors.R = (byte)r;
            quad.BottomLeft.Colors.G = (byte)g;
            quad.BottomLeft.Colors.B = (byte)b;
            quad.BottomLeft.Colors.A = (byte)a;
            quad.TopLeft.Colors.R = (byte)r;
            quad.TopLeft.Colors.G = (byte)g;
            quad.TopLeft.Colors.B = (byte)b;
            quad.TopLeft.Colors.A = (byte)a;
            quad.TopRight.Colors.R = (byte)r;
            quad.TopRight.Colors.G = (byte)g;
            quad.TopRight.Colors.B = (byte)b;
            quad.TopRight.Colors.A = (byte)a;
            quad.BottomRight.Colors.R = (byte)r;
            quad.BottomRight.Colors.G = (byte)g;
            quad.BottomRight.Colors.B = (byte)b;
            quad.BottomRight.Colors.A = (byte)a;

            quad.BottomLeft.Vertices.X = vertices[RegionAttachment.X1];
            quad.BottomLeft.Vertices.Y = vertices[RegionAttachment.Y1];
            quad.TopLeft.Vertices.X = vertices[RegionAttachment.X2];
            quad.TopLeft.Vertices.Y = vertices[RegionAttachment.Y2];
            quad.TopRight.Vertices.X = vertices[RegionAttachment.X3];
            quad.TopRight.Vertices.Y = vertices[RegionAttachment.Y3];
            quad.BottomRight.Vertices.X = vertices[RegionAttachment.X4];
            quad.BottomRight.Vertices.Y = vertices[RegionAttachment.Y4];

            quad.BottomLeft.TexCoords.U = self.UVs[RegionAttachment.X1];
            quad.BottomLeft.TexCoords.V = self.UVs[RegionAttachment.Y1];
            quad.TopLeft.TexCoords.U = self.UVs[RegionAttachment.X2];
            quad.TopLeft.TexCoords.V = self.UVs[RegionAttachment.Y2];
            quad.TopRight.TexCoords.U = self.UVs[RegionAttachment.X3];
            quad.TopRight.TexCoords.V = self.UVs[RegionAttachment.Y3];
            quad.BottomRight.TexCoords.U = self.UVs[RegionAttachment.X4];
            quad.BottomRight.TexCoords.V = self.UVs[RegionAttachment.Y4];

        }

        void _spAtlasPage_createTexture(AtlasPage self, string path)
        {

            CCTexture2D texture = CCTextureCache.SharedTextureCache.AddImage(path);
            CCTextureAtlas textureAtlas = new CCTextureAtlas();
            textureAtlas.InitWithTexture(texture, 128);

            textureAtlas.DrawQuads(); //retain();

            self.rendererObject = textureAtlas;
            self.width = texture.PixelsWide;
            self.height = texture.PixelsHigh;
        }

        void _spAtlasPage_disposeTexture(AtlasPage self)
        {
            ((CCTextureAtlas)self.rendererObject).RemoveAllQuads();//Release(); // =============> ?
        }

        #endregion

     

    }
}

File: spine-cocos2dx-v3/src/spine/CCSkeleton.cpp
/******************************************************************************
 * Spine Runtime Software License - Version 1.0
 * 
 * Copyright (c) 2013, Esoteric Software
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms in whole or in part, with
 * or without modification, are permitted provided that the following conditions
 * are met:
 * 
 * 1. A Spine Single User License or Spine Professional License must be
 *    purchased from Esoteric Software and the license must remain valid:
 *    http://esotericsoftware.com/
 * 2. Redistributions of source code must retain this license, which is the
 *    above copyright notice, this declaration of conditions and the following
 *    disclaimer.
 * 3. Redistributions in binary form must reproduce this license, which is the
 *    above copyright notice, this declaration of conditions and the following
 *    disclaimer, in the documentation and/or other materials provided with the
 *    distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *****************************************************************************/

#include <spine/CCSkeleton.h>
#include <spine/spine-cocos2dx.h>

USING_NS_CC;
using std::min;
using std::max;

namespace spine {

CCSkeleton* CCSkeleton::createWithData (spSkeletonData* skeletonData, bool ownsSkeletonData) {
	CCSkeleton* node = new CCSkeleton(skeletonData, ownsSkeletonData);
	node->autorelease();
	return node;
}

CCSkeleton* CCSkeleton::createWithFile (const char* skeletonDataFile, spAtlas* atlas, float scale) {
	CCSkeleton* node = new CCSkeleton(skeletonDataFile, atlas, scale);
	node->autorelease();
	return node;
}

CCSkeleton* CCSkeleton::createWithFile (const char* skeletonDataFile, const char* atlasFile, float scale) {
	CCSkeleton* node = new CCSkeleton(skeletonDataFile, atlasFile, scale);
	node->autorelease();
	return node;
}

void CCSkeleton::initialize () {
	atlas = 0;
	debugSlots = false;
	debugBones = false;
	timeScale = 1;

	blendFunc.src = GL_ONE;
	blendFunc.dst = GL_ONE_MINUS_SRC_ALPHA;
	setOpacityModifyRGB(true);

	setShaderProgram(CCShaderCache::getInstance()->getProgram(GLProgram::SHADER_NAME_POSITION_TEXTURE_COLOR));
	scheduleUpdate();
}

void CCSkeleton::setSkeletonData (spSkeletonData *skeletonData, bool ownsSkeletonData) {
	skeleton = spSkeleton_create(skeletonData);
	rootBone = skeleton->bones[0];
	this->ownsSkeletonData = ownsSkeletonData;	
}

CCSkeleton::CCSkeleton () {
	initialize();
}

CCSkeleton::CCSkeleton (spSkeletonData *skeletonData, bool ownsSkeletonData) {
	initialize();

	setSkeletonData(skeletonData, ownsSkeletonData);
}

CCSkeleton::CCSkeleton (const char* skeletonDataFile, spAtlas* atlas, float scale) {
	initialize();

	spSkeletonJson* json = spSkeletonJson_create(atlas);
	json->scale = scale;
	spSkeletonData* skeletonData = spSkeletonJson_readSkeletonDataFile(json, skeletonDataFile);
	CCAssert(skeletonData, json->error ? json->error : "Error reading skeleton data.");
	spSkeletonJson_dispose(json);

	setSkeletonData(skeletonData, true);
}

CCSkeleton::CCSkeleton (const char* skeletonDataFile, const char* atlasFile, float scale) {
	initialize();

	atlas = spAtlas_readAtlasFile(atlasFile);
	CCAssert(atlas, "Error reading atlas file.");

	spSkeletonJson* json = spSkeletonJson_create(atlas);
	json->scale = scale;
	spSkeletonData* skeletonData = spSkeletonJson_readSkeletonDataFile(json, skeletonDataFile);
	CCAssert(skeletonData, json->error ? json->error : "Error reading skeleton data file.");
	spSkeletonJson_dispose(json);

	setSkeletonData(skeletonData, true);
}

CCSkeleton::~CCSkeleton () {
	if (ownsSkeletonData) spSkeletonData_dispose(skeleton->data);
	if (atlas) spAtlas_dispose(atlas);
	spSkeleton_dispose(skeleton);
}

void CCSkeleton::update (float deltaTime) {
	spSkeleton_update(skeleton, deltaTime * timeScale);
}

void CCSkeleton::draw () {
	CC_NODE_DRAW_SETUP();

	GL::blendFunc(blendFunc.src, blendFunc.dst);
	Color3B color = getColor();
	skeleton->r = color.r / (float)255;
	skeleton->g = color.g / (float)255;
	skeleton->b = color.b / (float)255;
	skeleton->a = getOpacity() / (float)255;
	if (premultipliedAlpha) {
		skeleton->r *= skeleton->a;
		skeleton->g *= skeleton->a;
		skeleton->b *= skeleton->a;
	}

    int additive = 0;
	TextureAtlas* textureAtlas = 0;
	V3F_C4B_T2F_Quad quad;
	quad.tl.vertices.z = 0;
	quad.tr.vertices.z = 0;
	quad.bl.vertices.z = 0;
	quad.br.vertices.z = 0;
	for (int i = 0, n = skeleton->slotCount; i < n; i++) {
		spSlot* slot = skeleton->drawOrder[i];
		if (!slot->attachment || slot->attachment->type != ATTACHMENT_REGION) continue;
		spRegionAttachment* attachment = (spRegionAttachment*)slot->attachment;
        TextureAtlas* regionTextureAtlas = getTextureAtlas(attachment);

        if (slot->data->additiveBlending != additive) {
            if (textureAtlas) {
                textureAtlas->drawQuads();
                textureAtlas->removeAllQuads();
            }
            additive = !additive;
            GL::blendFunc(blendFunc.src, additive ? GL_ONE : blendFunc.dst);
        } else if (regionTextureAtlas != textureAtlas && textureAtlas) {
            textureAtlas->drawQuads();
            textureAtlas->removeAllQuads();
        }
        textureAtlas = regionTextureAtlas;

		int quadCount = textureAtlas->getTotalQuads();
		if (textureAtlas->getCapacity() == quadCount) {
			textureAtlas->drawQuads();
			textureAtlas->removeAllQuads();
			if (!textureAtlas->resizeCapacity(textureAtlas->getCapacity() * 2)) return;
		}

		spRegionAttachment_updateQuad(attachment, slot, &quad, premultipliedAlpha);
		textureAtlas->updateQuad(&quad, quadCount);
	}
	if (textureAtlas) {
		textureAtlas->drawQuads();
		textureAtlas->removeAllQuads();
	}

	if (debugSlots) {
		// Slots.
		DrawPrimitives::setDrawColor4B(0, 0, 255, 255);
		glLineWidth(1);
		Point points[4];
		V3F_C4B_T2F_Quad quad;
		for (int i = 0, n = skeleton->slotCount; i < n; i++) {
			spSlot* slot = skeleton->drawOrder[i];
			if (!slot->attachment || slot->attachment->type != ATTACHMENT_REGION) continue;
			spRegionAttachment* attachment = (spRegionAttachment*)slot->attachment;
			spRegionAttachment_updateQuad(attachment, slot, &quad);
			points[0] = Point(quad.bl.vertices.x, quad.bl.vertices.y);
			points[1] = Point(quad.br.vertices.x, quad.br.vertices.y);
			points[2] = Point(quad.tr.vertices.x, quad.tr.vertices.y);
			points[3] = Point(quad.tl.vertices.x, quad.tl.vertices.y);
			DrawPrimitives::drawPoly(points, 4, true);
		}
	}
	if (debugBones) {
		// Bone lengths.
		glLineWidth(2);
		DrawPrimitives::setDrawColor4B(255, 0, 0, 255);
		for (int i = 0, n = skeleton->boneCount; i < n; i++) {
			spBone *bone = skeleton->bones[i];
			float x = bone->data->length * bone->m00 + bone->worldX;
			float y = bone->data->length * bone->m10 + bone->worldY;
            DrawPrimitives::drawLine(Point(bone->worldX, bone->worldY), Point(x, y));
		}
		// Bone origins.
		DrawPrimitives::setPointSize(4);
		DrawPrimitives::setDrawColor4B(0, 0, 255, 255); // Root bone is blue.
		for (int i = 0, n = skeleton->boneCount; i < n; i++) {
			spBone *bone = skeleton->bones[i];
            DrawPrimitives::drawPoint(Point(bone->worldX, bone->worldY));
			if (i == 0) DrawPrimitives::setDrawColor4B(0, 255, 0, 255);
		}
	}
}

TextureAtlas* CCSkeleton::getTextureAtlas (spRegionAttachment* regionAttachment) const {
	return (TextureAtlas*)((spAtlasRegion*)regionAttachment->rendererObject)->page->rendererObject;
}

Rect CCSkeleton::getBoundingBox() const {
	float minX = FLT_MAX, minY = FLT_MAX, maxX = FLT_MIN, maxY = FLT_MIN;
	float scaleX = getScaleX();
	float scaleY = getScaleY();
	float vertices[8];
	for (int i = 0; i < skeleton->slotCount; ++i) {
		spSlot* slot = skeleton->slots[i];
		if (!slot->attachment || slot->attachment->type != ATTACHMENT_REGION) continue;
		spRegionAttachment* attachment = (spRegionAttachment*)slot->attachment;
		spRegionAttachment_computeWorldVertices(attachment, slot->skeleton->x, slot->skeleton->y, slot->bone, vertices);
		minX = min(minX, vertices[VERTEX_X1] * scaleX);
		minY = min(minY, vertices[VERTEX_Y1] * scaleY);
		maxX = max(maxX, vertices[VERTEX_X1] * scaleX);
		maxY = max(maxY, vertices[VERTEX_Y1] * scaleY);
		minX = min(minX, vertices[VERTEX_X4] * scaleX);
		minY = min(minY, vertices[VERTEX_Y4] * scaleY);
		maxX = max(maxX, vertices[VERTEX_X4] * scaleX);
		maxY = max(maxY, vertices[VERTEX_Y4] * scaleY);
		minX = min(minX, vertices[VERTEX_X2] * scaleX);
		minY = min(minY, vertices[VERTEX_Y2] * scaleY);
		maxX = max(maxX, vertices[VERTEX_X2] * scaleX);
		maxY = max(maxY, vertices[VERTEX_Y2] * scaleY);
		minX = min(minX, vertices[VERTEX_X3] * scaleX);
		minY = min(minY, vertices[VERTEX_Y3] * scaleY);
		maxX = max(maxX, vertices[VERTEX_X3] * scaleX);
		maxY = max(maxY, vertices[VERTEX_Y3] * scaleY);
	}
	Point position = getPosition();
	return Rect(position.x + minX, position.y + minY, maxX - minX, maxY - minY);
}

// --- Convenience methods for Skeleton_* functions.

void CCSkeleton::updateWorldTransform () {
	spSkeleton_updateWorldTransform(skeleton);
}

void CCSkeleton::setToSetupPose () {
	spSkeleton_setToSetupPose(skeleton);
}
void CCSkeleton::setBonesToSetupPose () {
	spSkeleton_setBonesToSetupPose(skeleton);
}
void CCSkeleton::setSlotsToSetupPose () {
	spSkeleton_setSlotsToSetupPose(skeleton);
}

spBone* CCSkeleton::findBone (const char* boneName) const {
	return spSkeleton_findBone(skeleton, boneName);
}

spSlot* CCSkeleton::findSlot (const char* slotName) const {
	return spSkeleton_findSlot(skeleton, slotName);
}

bool CCSkeleton::setSkin (const char* skinName) {
	return spSkeleton_setSkinByName(skeleton, skinName) ? true : false;
}

spAttachment* CCSkeleton::getAttachment (const char* slotName, const char* attachmentName) const {
	return spSkeleton_getAttachmentForSlotName(skeleton, slotName, attachmentName);
}
bool CCSkeleton::setAttachment (const char* slotName, const char* attachmentName) {
	return spSkeleton_setAttachment(skeleton, slotName, attachmentName) ? true : false;
}

// --- BlendProtocol
    
const BlendFunc& CCSkeleton::getBlendFunc() const {
    return blendFunc;
}
    
void CCSkeleton::setBlendFunc( const BlendFunc &blendFunc) {
    this->blendFunc = blendFunc;
}
    
void CCSkeleton::setOpacityModifyRGB (bool value) {
    premultipliedAlpha = value;
}
    
bool CCSkeleton::isOpacityModifyRGB () const {
    return premultipliedAlpha;
}

}
